ingest/src/main/java/com/microsoft/azure/kusto/ingest/IngestionProperties.java

```diff
@@ -303,27 +310,39 @@ void validate() throws IngestionClientException {
         IngestionMapping.IngestionMappingKind ingestionMappingKind = ingestionMapping.getIngestionMappingKind();
         String message = null;
 
-        if (ingestionMapping.getColumnMappings() != null) {
-            if (StringUtils.isNotBlank(mappingReference)) {
-                message += String.format("Both mapping reference '%s' and column mappings were defined.", mappingReference);
-            }
-            for (ColumnMapping column : ingestionMapping.getColumnMappings()) {
-                if (!column.isValid(ingestionMappingKind)) {
-                    message += String.format("Column mapping '%s' is invalid", column.columnName);
-                }
-            }
-        } else if (StringUtils.isBlank(mappingReference)) {
+        if ((ingestionMapping.getColumnMappings() == null) && StringUtils.isBlank(mappingReference)) {
             if (dataFormat.isMappingRequired()) {
                 message += String.format("Mapping must be specified for '%s' format.", dataFormat.name());
             }
 
-            if (ingestionMappingKind != null && ingestionMappingKind != IngestionMapping.IngestionMappingKind.unknown) {
-                message += String.format("If IngestionMappingKind was defined ('%s'), mapping must be specified.", ingestionMappingKind);
+            if (ingestionMappingKind != null) {
+                message += String.format("IngestionMappingKind was defined ('%s'), so a mapping must be defined as well.", ingestionMappingKind);
             }
-        }
+        } else { // a mapping was provided
+            if (dataFormat.getIngestionMappingKind() == null) {
+                message += String.format("Data format '%s' does not support mappings", dataFormat.name());
+            } else if (!dataFormat.getIngestionMappingKind().equals(ingestionMappingKind)) {
+                message += String.format("Wrong ingestion mapping for format '%s'; mapping kind should be '%s', but was '%s'.",
+                        dataFormat.name(), dataFormat.getIngestionMappingKind(), ingestionMappingKind != null ? ingestionMappingKind.name() : "null");
+            }
+
+            if (ingestionMapping.getColumnMappings() != null) {
+                if (StringUtils.isNotBlank(mappingReference)) {
+                    message += String.format("Both mapping reference '%s' and column mappings were defined.", mappingReference);
+                }
 
-        if (dataFormat.isMappingRequired() && !dataFormat.getIngestionMappingKind().equals(ingestionMappingKind)) {
-            message += String.format("Wrong ingestion mapping for format '%s'; found '%s' mapping kind.", dataFormat.name(), ingestionMappingKind != null ? ingestionMappingKind.name() : "null");
+                if (ingestionMappingKind == null) {
+                    message += "Column mappings were provided, so the IngestionMappingKind must be provided as well.";
```

Comment:
In the Java SDK, setting either type of mapping requires also passing an `IngestionMappingKind`

Reply:
I'm debating whether to remove this now, or in 3.0.1. I think it's best to wait.

Reply:
Nevermind, this check is new in this PR, so I'll just remove it now.